### ISSUE TITLE
Discover secrets in base64 encoded strings

### ISSUE THREAD

**Issue by oleyv-cl** [gitleaks/gitleaks#807]
**Is your feature request related to a problem? Please describe.**
No. 

**Describe the solution you'd like**
It is currently possible to search for base64 encoded string. After the base64 sting is found, would it be possible to decode the string and pass it through the gitleaks detection logic again in the decoded form to discover potential secrets within base64?

**Describe alternatives you've considered**
Get the base64 string to stdout; pipe the output to base64 -d; pipe the output to gitleaks --no-git. 

**Additional context**
Thank you very much for a great and very flexible tool. 

cc @zricethezav

**Comment by zricethezav** [comment]
> Get the base64 string to stdout; pipe the output to base64 -d; pipe the output to gitleaks --no-git.

This feature request sounds more like you want me to add support for accepting piped input, is that correct?

**Comment by OleyV** [comment]
Re-reading the request I've realized that piped input will definitely help without overcomplicating the code logic.

**Comment by sergiomarotco** [comment]
I vote for this feature

**Comment by leite08** [comment]
It would be super useful for our use case to be able to run a custom configuration against the result of a decoded base64 string.

I'm new to this project so I don't know whether this would make sense to how the underlying engine works, but by looking at the default config it seems we could have a "decodeBase64" flag to indicate the respective config/rule's regex should be applied at the output of base64 decode.

To make it more flexible, instead of a flag we could have a rule setting called `pre-processor` where one would indicate which pre-processor to apply to that rule from a list of supported pre-processors (one being "base64Decoder").

Happy to create a new issue if this comment doesn't apply to the current issue.

**Comment by rgmz** [comment]
> It would be super useful for our use case to be able to run a custom configuration against the result of a decoded base64 string.
> To make it more flexible, instead of a flag we could have a rule setting called pre-processor where one would indicate which pre-processor to apply to that rule from a list of supported pre-processors (one being "base64Decoder").

Pre-processors have many potential use cases — such as decoding base64 (as mentioned), decoding JWTs, or deflating archives — but would the feature would need to be carefully designed and implemented to preserve performance.

@zricethezav what are your thoughts on adding a net-new feature like this? 

----

It's worth noting that [TruffleHog](https://github.com/trufflesecurity/trufflehog) has built-in decoders for Base64 and archives, plus rules are written programmatically (rather than in TOML) which allows for additional flexibility and functionality.

**Comment by imre-uncinc** [comment]
base64 scanning would be a useful feature, we've had an incident with a base64-encoded secret in a repository before at my place of work.

**Comment by bplaxco** [comment]
I'm probably going to implement something like base64 decoding for [this scanner project that leverages Gitleaks](https://github.com/leaktk/scanner). I played around with some ideas from Trufflehog and managed to get a little bit better speed for in-place decoding but that wasn't keeping track of where the original value was in the text.

I'm thinking the feature will:

- decode it in-place
- pass the fragment with decoded content to the detector
- report the position of the base64 blob for the line number info reported in the results

If the performance is good, do you want me to drop a note in this request and/or file a MR to have it pulled into Gitleaks itself?


Side note: The workaround I'm using today is for things that are really likely to be base64 encoded (e.g. [certain JWTs](https://github.com/leaktk/patterns/blob/c6017d638412636308d155e6144226f8c91d2afd/patterns/gitleaks/8.18.2/98-general.toml#L3)) is to use this `gen-base64-search-regex` script:

```python
#! /usr/bin/python3
import sys
import base64
import math

USAGE = """
Genrate a regex to search base64 encoded versions of a string.

Note this can generate regular base64 (the default) or URL-safe base64

USAGE

    gen-base64-search-regex [OPTIONS] <search-string>


OPTIONS

    --url   use URL-safe base64 (default to regular base64)
    --help  print this text

EXAMPLE

    $ gen-base64-search-regex --url '"sub":"system:serviceaccount:'
    > (InN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudD|JzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6|ic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50O)
"""


def encode(target, method, pad=0):
    """
    This does some permutations of the target with different levels of padding
    and then chops off the characters that would be impacted by any surrounding
    text.

    The math:
        8 comes from the number of bits in a byte
        6 comes from the number of bits in a b64 character

        This basically looks at the number of bytes that would change because
        of the prefix or suffix to the data and chopts them off in the b64
        encoded data.
    """
    padding = b"x" * pad
    data = padding + target.encode("UTF-8")
    start = math.ceil(pad * 8 / 6)
    end = math.floor(len(data) * 8 / 6)

    if start == end:
        # You would end up with empty values that match everything
        raise ValueError("target too small to generate regex")

    return method(data)[start:end].decode()


def main():
    target = None
    flags = {}

    for arg in sys.argv[1:]:
        if arg == "--help":
            flags["help"] = True
        elif arg == "--url":
            flags["url"] = True
        else:
            target = arg

    if flags.get("help") or not target:
        print(USAGE)
        sys.exit(1)

    if flags.get("url"):
        method = base64.urlsafe_b64encode
    else:
        method = base64.b64encode

    patterns = (encode(target, method, pad=pad) for pad in range(0, 3))

    print(f"({'|'.join(patterns)})")


if __name__ == "__main__":
    main()
```

**Comment by zricethezav** [comment]
@bplaxco I'd be happy to introduce a b64 decoder similar to TruffleHog. Not sure I have the time right now to introduce that feature but would happily review a PR. 

> I'm thinking the feature will:
> decode it in-place
> pass the fragment with decoded content to the detector
> report the position of the base64 blob for the line number info reported in the results

I like this as a start but I think it would be better to support decoding b64 in parallel with plaintext scans, in other words gitleaks would scan the plaintext fragment _and_ the b64 decoded fragment. We would have to work on deduplication logic as well since there could be a case where a plaintext secret and b64 decoded secret both satisfy a rule.

This feature has been long overdue tbh :/

**Comment by bplaxco** [comment]
Okay, yeah I can do something like the pseudo code below:

```go
if B64Detected(fragment) {
    results = append(results, detector.Detect(DecodeInPlace(fragment))**)
}

results = append(results, detector.Detect(fragment)**)
```

Should add a few milliseconds of extra decoding + the time to do the detection if there's b64 based on some performance testing I did before.

If that sounds good, I'll take a crack at it in my scanner, and if you like it, I'll send a MR your way.

**Comment by zricethezav** [comment]
@bplaxco that looks like a great start!

**Comment by bplaxco** [comment]
@zricethezav I've started some work over in https://github.com/gitleaks/gitleaks/pull/1488

I still need to test it more and clean up and make sure there's no performance hits, but what do you think of the overall approach so far?

**Comment by KraMorK** [comment]
The function is very necessary. can we consider the option of adding the function activation as a separate startup parameter, if we are worried about performance, so as not to affect other users?

**Comment by bplaxco** [comment]
Yeah, that should be pretty straight forward to do.

The last thing I think I have to do before I can call it ready (apart from adding a flag to enable it optionally) is finish some finding location mapping since things shift around a lot as you decode them.

**Comment by bplaxco** [comment]
Update on https://github.com/gitleaks/gitleaks/pull/1488

Working Now:

- Finds strings in base64 and decodes them (this even works if it's encoded again with surrounding text (e.g. a jwt with b64 encoded secret in the payload)
- Avoids returning findings from previous passes when it does it's decoding pass.
- Handles getting the location pretty right (it currently reports the bounds of the encoded item for the match)
- Handles the contents of the file shifting around as strings change size due to decoding
- Tries to bail out early to avoid extra work 
- Unit tests exist for the new changes

So it's functional at this point.

Still TODO though:

- Set up a CLI flag to enable it (so that it can be off by default. I might just call this `--max-decode-depth` and it defaults to zero so that you can determine how deep down the rabbit hole you want to go with decoding multi encoded values. And a depth of 0 is the same as off).
- See if I can have the location info properly handle cases where the match starts outside the bounds of the encoded item (I think I know how I can do that)
- Performance test it
  - If it slows down on big projects, I may add some additional bail early checks to avoid having to try to decode as many things. (e.g. https://github.com/gitleaks/gitleaks/pull/1488#discussion_r1733476685)

@rgmz, @zricethezav && anyone else here up for running the current set of changes on my branch against a few things to see if it's reporting like you'd expect

**Comment by rgmz** [comment]
> Finds strings in base64 and decodes them (this even works if it's encoded again with surrounding text (e.g. a jwt with b64 encoded secret in the payload)

Does it do recursive base64 decoding? I've found a few tricky secrets with `n > 1` nesting.

**Comment by bplaxco** [comment]
Yep! It does recursive decoding and keeps up with the location in the original text 👍

I'm going to make the depth a cli flag, `--max-decode-depth`, that defaults to zero (effectively disabling the feature unless you want it).

**Comment by bplaxco** [comment]
Proper location handling is in place :tada: ([code](https://github.com/gitleaks/gitleaks/pull/1488/commits/799ed9794d7c7b8b0533f607a7ddf0b81a23115b)) ([tests](https://github.com/gitleaks/gitleaks/pull/1488/commits/75fd0158589add32c7c1e694e1053d0c99ccd098))

Behavior:

- If at any point up the tree of encoded segments it's fully within the bounds of an encoded segment, the bounds of its "root" encoded segment (e.g. the encoded segment in the original text) are returned as the matchIndex

- If it goes outside the bounds of the root encoded segment, the match index is adjusted to handle that. Something kinda like:
  ```
  adjustedStart = min(matchIndexInOriginal, rootEncodedStart)
  adjustedEnd = max(matchIndexInOriginal, rootEncodedEnd)
  ```

It also handles all the fun weird recursive cases like it going outside of one segment but not its parent, etc. And shortcuts any recursion if it's cleanly inside a segment up the tree.

Next up CLI flag to set the max decode depth. :+1:

**Comment by bplaxco** [comment]
CLI flag added and rebased on the latest changes in master :tada: 

I think all that's left is to:

- Benchmark it against what's currently on master and stress test it—making adjustments if there's any performance issues
- Run it against some stuff I've seen in the past to see how it does and some other ad-hoc cases to make sure things are solid and working as expected (I've done some of this alredy)
- Make any changes to code or docs requested during review

@zricethezav && @rgmz, if y'all can think of any particularly gnarly cases to run it through, please let me know ^_^

**Comment by zricethezav** [comment]
> Yep! It does recursive decoding and keeps up with the location in the original text 👍 

🤯 dude this shit is wild. Amazing work. I'll try to give it an in-depth review this weekend

**Comment by bplaxco** [comment]
Awesome, thanks! Feel free to nit pick the heck out of it or come up with any edge cases you'd like me to check. 👍

**Comment by zricethezav** [comment]
@bplaxco is the PR ready for review or is it still draft?

**Comment by bplaxco** [comment]
@zricethezav I'm planning on stress testing it today and making a few performance tweaks if needed and then hopefully marking it ready for review.

If it's still marked as a draft by the weekend, then it'd probably be in a good enough state for a high level review to make sure there's no big things you disagree with. But at the moment it looks like it's working as expected.

**Comment by bplaxco** [comment]
So far the performance is looking good on small runs (actually a bit better than what's on master—even with decoding enabled, probably due to moving the stop words around but that's only a guess).

I'm putting it through it's paces now, but I may have to kill it before it's done and kick it off again when I get home.

I'm also doing some testing to see if there are some [things I can wholesale skip](https://github.com/gitleaks/gitleaks/pull/1488#discussion_r1779131890) to speed it up more, and then I'll do that last test case that's commented out.

And at first glance it looks like they're both producing the same results:

```
$ ./bin/b64d-gitleaks detect --source repos/fake-leaks --config etc/leaktk-patterns.toml --report-path /dev/stdout 2> /dev/null | jq '.[].Fingerprint' -r | sort | sha256sum 
3a8bfa661aa5edc552ee8a65221ea1212318cc386475525641afc134e7879474  -

$ ./bin/reg-gitleaks detect --source repos/fake-leaks --config etc/leaktk-patterns.toml --report-path /dev/stdout 2> /dev/null | jq '.[].Fingerprint' -r | sort | sha256sum 
3a8bfa661aa5edc552ee8a65221ea1212318cc386475525641afc134e7879474  -
```

Current perf test that I'm doing if anyone's interested:

```sh
#! /usr/bin/env bash

if [[ ! -f etc/leaktk-patterns.toml ]]
then
  echo 'Fetching leaktk patterns'
  mkdir -p etc
  curl -o etc/leaktk-patterns.toml https://raw.githubusercontent.com/leaktk/patterns/cb9c17866d931d972e5c613fe67a74b172ddd40b/target/patterns/gitleaks/8.18.2 &> /dev/null
fi

if [[ ! -d src ]]
then
  echo "Fetching source code"
  git clone https://github.com/leaktk/gitleaks src
fi

mkdir -p bin

if [[ ! -f bin/b64d-gitleaks ]]
then
  echo "Building bin/b64d-gitleaks"
  (
    cd src
    git checkout 6d6c5d5c08872a5d4aff2c6d6f51a525da6e575d # decode-base64 (no optimizations)
    git -dfX
    make build
    mv ./gitleaks ../bin/b64d-gitleaks
  )
fi

if [[ ! -f bin/reg-gitleaks ]]
then
  echo "Building bin/reg-gitleaks"
  (
    cd src
    git checkout ed19c4e8b417313e7ab6878327167c558f5f98f8 # master
    git -dfX
    make build
    mv ./gitleaks ../bin/reg-gitleaks
  )
fi

if [[ ! -d repos ]]
then
  mkdir -p repos
  echo "Fetching Repos"
  git clone https://github.com/clojure-emacs/cider repos/cider
  git clone https://github.com/leaktk/fake-leaks repos/fake-leaks
  git clone https://github.com/kubernetes/kubernetes repos/kubernetes
  git clone https://github.com/torvalds/linux repos/linux
  git clone https://github.com/nadimkobeissi/mkbsd repos/mkbsd
  git clone https://github.com/WinampDesktop/winamp repos/winamp
fi

function scan-all {
  scanner="$1"
  scan_id="$2"

  for repo in `ls repos`
  do
    run_scan="./bin/${scanner} detect --source repos/${repo} --report-path reports/${scan_id}-${repo}.json"
    echo ">> Running: ${run_scan}"
    ${run_scan} &> /dev/null
  done
  echo ">> Times:"
}

function test-scanner {
  scanner="$1"
  scan_id="$(echo "${scanner}" | sha256sum | cut -d ' ' -f 1 | cut -c 1-8)"
  echo "======================================================================"
  echo "Results (id=$scan_id cmd='${scanner}'):"
  echo
  echo "> LeakTK Patterns:"
  time scan-all "${scanner} --config etc/leaktk-patterns.toml" "${scan_id}"
  echo
  echo "> Default Patterns:"
  time scan-all "$1"
  echo "======================================================================"
  echo
}

rm -rf reports; mkdir -p reports

# Baseline
test-scanner reg-gitleaks

# Changes with decoding disabled
test-scanner b64d-gitleaks

# Changes with decoding enabled
test-scanner 'b64d-gitleaks --max-decode-depth=8'

# TODO Changes with decoding enabled and performance improvments
# test-scanner 'perf-b64d-gitleaks --max-decode-depth=8'
```

**Comment by bplaxco** [comment]
I think https://github.com/gitleaks/gitleaks/pull/1488 is ready for review. The performance tests are looking good but I'm letting another round complete and then I'll post the results here ^_^

**Comment by bplaxco** [comment]
Results from running three versions: 

* `reg-gitleaks` - the code before my changes
* `b64d-gitleaks` - the initial round of changes
* `perf-b64d-gitleaks` - some slight tweaks made to skip obvious things like method names and such

I ran the tests twice here just for the sake of seeing how it does.

Note the flags: I run the code with and without the `--max-decode-depth` set since by default it's off. 

```
======================================================================
Results (id=b28700b3 cmd='reg-gitleaks'):
>> Running: ./bin/reg-gitleaks detect --source repos/cider --report-path reports/b28700b3-cider.json
>> Running: ./bin/reg-gitleaks detect --source repos/fake-leaks --report-path reports/b28700b3-fake-leaks.json
>> Running: ./bin/reg-gitleaks detect --source repos/kubernetes --report-path reports/b28700b3-kubernetes.json
>> Running: ./bin/reg-gitleaks detect --source repos/linux --report-path reports/b28700b3-linux.json
>> Running: ./bin/reg-gitleaks detect --source repos/mkbsd --report-path reports/b28700b3-mkbsd.json
>> Running: ./bin/reg-gitleaks detect --source repos/winamp --report-path reports/b28700b3-winamp.json
>> Times:

real	53m40.069s
user	141m22.271s
sys	3m21.016s
======================================================================

======================================================================
Results (id=68fb7248 cmd='b64d-gitleaks'):
>> Running: ./bin/b64d-gitleaks detect --source repos/cider --report-path reports/68fb7248-cider.json
>> Running: ./bin/b64d-gitleaks detect --source repos/fake-leaks --report-path reports/68fb7248-fake-leaks.json
>> Running: ./bin/b64d-gitleaks detect --source repos/kubernetes --report-path reports/68fb7248-kubernetes.json
>> Running: ./bin/b64d-gitleaks detect --source repos/linux --report-path reports/68fb7248-linux.json
>> Running: ./bin/b64d-gitleaks detect --source repos/mkbsd --report-path reports/68fb7248-mkbsd.json
>> Running: ./bin/b64d-gitleaks detect --source repos/winamp --report-path reports/68fb7248-winamp.json
>> Times:

real	53m20.830s
user	138m51.497s
sys	3m15.884s
======================================================================

======================================================================
Results (id=3d4750d3 cmd='b64d-gitleaks --max-decode-depth=8'):
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/cider --report-path reports/3d4750d3-cider.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/fake-leaks --report-path reports/3d4750d3-fake-leaks.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/kubernetes --report-path reports/3d4750d3-kubernetes.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/linux --report-path reports/3d4750d3-linux.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/mkbsd --report-path reports/3d4750d3-mkbsd.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/winamp --report-path reports/3d4750d3-winamp.json
>> Times:

real	54m11.879s
user	160m28.833s
sys	3m16.377s
======================================================================

======================================================================
Results (id=e59b2021 cmd='perf-b64d-gitleaks'):
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/cider --report-path reports/e59b2021-cider.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/fake-leaks --report-path reports/e59b2021-fake-leaks.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/kubernetes --report-path reports/e59b2021-kubernetes.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/linux --report-path reports/e59b2021-linux.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/mkbsd --report-path reports/e59b2021-mkbsd.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/winamp --report-path reports/e59b2021-winamp.json
>> Times:

real	52m11.501s
user	137m4.853s
sys	3m11.365s
======================================================================

======================================================================
Results (id=1cc5ab9c cmd='perf-b64d-gitleaks --max-decode-depth=8'):
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/cider --report-path reports/1cc5ab9c-cider.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/fake-leaks --report-path reports/1cc5ab9c-fake-leaks.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/kubernetes --report-path reports/1cc5ab9c-kubernetes.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/linux --report-path reports/1cc5ab9c-linux.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/mkbsd --report-path reports/1cc5ab9c-mkbsd.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/winamp --report-path reports/1cc5ab9c-winamp.json
>> Times:

real	54m25.868s
user	161m4.827s
sys	3m18.274s
======================================================================

======================================================================
Results (id=b28700b3 cmd='reg-gitleaks'):
>> Running: ./bin/reg-gitleaks detect --source repos/cider --report-path reports/b28700b3-cider.json
>> Running: ./bin/reg-gitleaks detect --source repos/fake-leaks --report-path reports/b28700b3-fake-leaks.json
>> Running: ./bin/reg-gitleaks detect --source repos/kubernetes --report-path reports/b28700b3-kubernetes.json
>> Running: ./bin/reg-gitleaks detect --source repos/linux --report-path reports/b28700b3-linux.json
>> Running: ./bin/reg-gitleaks detect --source repos/mkbsd --report-path reports/b28700b3-mkbsd.json
>> Running: ./bin/reg-gitleaks detect --source repos/winamp --report-path reports/b28700b3-winamp.json
>> Times:

real	52m59.735s
user	140m43.138s
sys	3m18.074s
======================================================================

======================================================================
Results (id=68fb7248 cmd='b64d-gitleaks'):
>> Running: ./bin/b64d-gitleaks detect --source repos/cider --report-path reports/68fb7248-cider.json
>> Running: ./bin/b64d-gitleaks detect --source repos/fake-leaks --report-path reports/68fb7248-fake-leaks.json
>> Running: ./bin/b64d-gitleaks detect --source repos/kubernetes --report-path reports/68fb7248-kubernetes.json
>> Running: ./bin/b64d-gitleaks detect --source repos/linux --report-path reports/68fb7248-linux.json
>> Running: ./bin/b64d-gitleaks detect --source repos/mkbsd --report-path reports/68fb7248-mkbsd.json
>> Running: ./bin/b64d-gitleaks detect --source repos/winamp --report-path reports/68fb7248-winamp.json
>> Times:

real	54m38.496s
user	142m26.405s
sys	3m23.131s
======================================================================

======================================================================
Results (id=3d4750d3 cmd='b64d-gitleaks --max-decode-depth=8'):
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/cider --report-path reports/3d4750d3-cider.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/fake-leaks --report-path reports/3d4750d3-fake-leaks.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/kubernetes --report-path reports/3d4750d3-kubernetes.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/linux --report-path reports/3d4750d3-linux.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/mkbsd --report-path reports/3d4750d3-mkbsd.json
>> Running: ./bin/b64d-gitleaks --max-decode-depth=8 detect --source repos/winamp --report-path reports/3d4750d3-winamp.json
>> Times:

real	54m43.539s
user	161m44.087s
sys	3m18.309s
======================================================================

======================================================================
Results (id=e59b2021 cmd='perf-b64d-gitleaks'):
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/cider --report-path reports/e59b2021-cider.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/fake-leaks --report-path reports/e59b2021-fake-leaks.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/kubernetes --report-path reports/e59b2021-kubernetes.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/linux --report-path reports/e59b2021-linux.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/mkbsd --report-path reports/e59b2021-mkbsd.json
>> Running: ./bin/perf-b64d-gitleaks detect --source repos/winamp --report-path reports/e59b2021-winamp.json
>> Times:

real	53m9.076s
user	138m46.731s
sys	3m13.913s
======================================================================

======================================================================
Results (id=1cc5ab9c cmd='perf-b64d-gitleaks --max-decode-depth=8'):
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/cider --report-path reports/1cc5ab9c-cider.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/fake-leaks --report-path reports/1cc5ab9c-fake-leaks.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/kubernetes --report-path reports/1cc5ab9c-kubernetes.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/linux --report-path reports/1cc5ab9c-linux.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/mkbsd --report-path reports/1cc5ab9c-mkbsd.json
>> Running: ./bin/perf-b64d-gitleaks --max-decode-depth=8 detect --source repos/winamp --report-path reports/1cc5ab9c-winamp.json
>> Times:

real	55m18.081s
user	162m49.592s
sys	3m20.716s
======================================================================
```

Here's making sure the results were the same (at least for the kubernetes results) when the `--max-decode-depth` flag isn't set so if this is merged in, it _shouldn't_ hurt anything by default if there's a bug in the logic.

```
# reg-gitleaks
$ cat reports/b28700b3-kubernetes.json | jq .[].Fingerprint -r | sort | sha256sum 
504bdd0895af01a296f18bb074698f47af71a85cf29e4eae3200a9ef1d7b8754

# b64d-gitleaks
$ cat reports/68fb7248-kubernetes.json | jq .[].Fingerprint -r | sort | sha256sum 
504bdd0895af01a296f18bb074698f47af71a85cf29e4eae3200a9ef1d7b8754  -

# perf-b64d-gitleaks
 $ cat reports/e59b2021-kubernetes.json | jq .[].Fingerprint -r | sort | sha256sum 
504bdd0895af01a296f18bb074698f47af71a85cf29e4eae3200a9ef1d7b8754  -
```

So it looks like the times are pretty comparable. And when I was running `b64d-gitleaks` and `perf-b64d-gitleaks` against smaller repos without the decoding on I was getting slightly faster times than `reg-gitleaks`.

I think I'm pretty happy with the performance. ^_^

@zricethezav the `perf-b64d-gitleaks` has a [check in it](https://github.com/gitleaks/gitleaks/pull/1488/files#r1779719212) that can rarely miss some base64 in it. More details in the PR. I'm not sure it makes too much of a difference if it's in there or not. What are your thoughts?

**Comment by zricethezav** [comment]
@bplaxco after doing a little further testing I believe there is a bug with printing when using `-v/--verbose`. Looks like the color output and vertical alignment breaks.
<img width="416" alt="Screenshot 2024-10-02 at 8 18 18 PM" src="https://github.com/user-attachments/assets/49b7f432-02f8-450d-8655-19ade505372f">

I'm testing with `gitleaks dir --max-decode-depth=8 -v`

**Comment by bplaxco** [comment]
@zricethezav would this help?

https://github.com/gitleaks/gitleaks/pull/1440

**Comment by zricethezav** [comment]
@bplaxco sure did! Thanks!
